design-docs/005-manifest-split.md

```diff
@@ -0,0 +1,406 @@
+        cardinality: null
+
+    - default:                  # default set must always exist, from defaults if needed
+                                # max-size: 1_000_000 by default (we'll have to tune)
+                                # default set is the only one that can have members that
+                                # go beyond the max-size
+                                # (in case of large arrays that don't fit)
+                                # default always has cardinality = null
+
+  rules:                        # what arrays go to each manifest set, an ordered list
+
+    # each rule has a target (the manifest set) and a set of conditions that are and-ed
+    # rules are applied in the order declared, and they break on the first match
+
+    - path: .*/(latitude|longitude|time)  # an optional regex matching on path
+
+      metadata-chunks: [0, 500]           # arrays having number of chunks in this range
+                                          # both extremes can be null
+                                          # in the future we'll implement initialized-chunks
+                                          # that will be useful for sparse arrays
+
+      target: coord1                      # arrays that match will go to this manifest set
+      
+
+    - metadata-chunks: [0, 200]           # arrays < 200 chunks will go to coord2
+                                          # but only if they didn't get assigned by the rule above
+      target: coord2
+
+
+    - metadata-chunks: [2000000, null]    # huge arrays have their special manifest set
+      target: big-array
+
+
+  preload:                      # what manifests to asynchronously preload on session start
+    max-manifest-size: 50000    # don't preload manifests larger than this
+    max-manifests: 10           # don't preload more than this many manifests
+    arrays:                     # currently we support selecting by path only
+                                # we will refine with more power if needed
+
+      - path: .*/time           # will preload manifests for all arrays matching this regex
+      - path: .*/latitude
+      - path: .*/longitude
+```
+
+The full power of the configuration is there for users that really need it. But
+usually the default configuration will be enough.
+
+Here is the default configuration:
+
+```yaml
+chunk-manifests:  # of course, we'll have to tune all these numbers
+  sets:
+    - coordinates:                # we intend coordinate arrays to be assigned here
+        max-manifest-size: 50000
+        cardinality: 1
+        max-arrays-per-manifest: null 
+        overflow-to: default
+
+    - default:
+        max-manifest-size: 1000000
+        cardinality: null
+        max-arrays-per-manifest: null
+        overflow-to: null
+
+  rules:
+    - metadata-chunks: [0, 5000]
+      path: .*
+      target: coordinates
+
+
+  # should we actually do this by default or is no-preload a better default?
+  preload:
```

Comment:
Actually, can we allow the user to preload the `coord1` set. Seems silly to duplicate info like the `path` regex in both `preload` and `rules`

Reply:
We could ... but currently the manifest sets are purely tacit. They don't persist in any way, it's purely a way to group them during flush. But of course, we could persist them. 

But, it feels less powerful? How about having a condition on manifest size too? so you could say "preload up to 5 small manifest" or "preload manifest for this array"

Reply:
Another future use-case is to preload the most recent manifest for multiple arrays (in the case of an updating forecast dataset)